src/chunking.ts

```diff
@@ -0,0 +1,180 @@
+
+    if (bytesUsed > maxBytes) {
+      // The last code point made us exceed the max size. We can't include it.
+      // The loop will terminate, and `i` will point to the start of this
+      // code point, effectively excluding it from the slice.
+      bytesUsed = before;
+      break;
+    }
+    // Advance the index by the number of 16-bit units this code point used.
+    i += codeUnits;
+  }
+
+  return i;
+}
+
+/**
+ * Converts a JSON object into an async generator of chunk frames.
+ *
+ * We use an async generator in order to allow other tasks to complete while
+ * the chunking is happening.
+ *
+ * Chunks the payload at character boundaries, and is aware of "wide" characters
+ * like 😀 (U+1F600) which is 4 UTF-8 bytes, but only 2 JS String#length
+ * characters. It never splits across a surrogate pair boundary (i.e., never
+ * splits a single Unicode code point).
+ *
+ * @param payload - the payload to be chunked
+ * @param chunkSize - the size of each chunk in bytes. If `0`, returns the
+ * payload as is. If the given `chunkSize` will result in a single frame, the
+ * payload is returned as is.
+ * @yields - a chunk frame or the original payload if it fits within a single
+ * frame
+ */
+
+export async function * toFrames<Payload extends Json>(
+  payload: Payload,
+  chunkSize: number,
+): AsyncGenerator<TransportChunkFrame | Payload, void> {
+  let begin = performance.now();
+  const json = JSON.stringify(payload);
+  const payloadLength = json.length;
+
+  // we need to leave space for our header
+  const id = getNextId();
+  let index = 0;
+  let seq = 0;
+
+  do {
+    if (performance.now() - begin >= FRAME_BUDGET_MS) {
+      // prevent background processing from locking up for too long by allowing
+      // other macro tasks to perform work between chunks
+      await tick();
+      begin = performance.now();
+    }
+    const header = `${id}|${seq}|` as const;
+    const headerSize = header.length + 1; // +1 for fin
+    const dataMaxLength = chunkSize - headerSize;
+
+    const remainingLen = payloadLength - index;
+    let end: number;
+
+    // Check if the rest of the string is guaranteed to fit.
+    // We use a conservative upper bound: assume each remaining UTF‑16 code unit
+    // costs 4 bytes in JSON.stringify(frame), plus 2 outer quotes for the whole
+    // frame. This overestimates (non‑BMP code points are 4 bytes but span 2
+    // code units; ASCII is 1 byte; quotes/backslashes are 2 bytes), so if this
+    // bound fits, the actual bytes will fit as well.
+    if (remainingLen * 3 + 2 <= dataMaxLength) {
+      // If the max possible size of the remainder fits, take it all and skip
+      // the very expensive `getJsonSliceIndex` call.
+      end = payloadLength;
+    } else {
+      // Otherwise, do the detailed calculation.
+      end = getJsonSliceIndex(json, index, dataMaxLength);
+    }
+
+    const fin = end === payloadLength ? 1 : 0;
+    const frame: TransportChunkFrame = `${header}${fin}${json.slice(
```

Comment:
I appreciate this is late in the process, but are we confident there no standard package that could encapsulate this for us, and efficiently convert JS objects to string based chunks?